modules/platforms/dotnet/Apache.Ignite/Internal/Compute/Executor/JobLoadContextCache.cs

```diff
@@ -0,0 +1,241 @@
+    {
+        await _cacheLock.WaitAsync().ConfigureAwait(false);
+
+        try
+        {
+            return GetOrAddImpl();
+        }
+        finally
+        {
+            _cacheLock.Release();
+        }
+
+        JobLoadContext GetOrAddImpl()
+        {
+            ref (JobLoadContext Ctx, long Ts) valRef = ref CollectionsMarshal.GetValueRefOrAddDefault(_jobLoadContextCache, paths, out var exists);
+
+            if (!exists)
+            {
+                valRef.Ctx = DeploymentUnitLoader.GetJobLoadContext(paths);
+
+                foreach (var path in paths.Paths)
+                {
+                    ref var listRef = ref CollectionsMarshal.GetValueRefOrAddDefault(_deploymentUnitSets, path, out _);
+                    listRef ??= new List<DeploymentUnitPaths>();
+
+                    listRef.Add(paths);
+                }
+            }
+
+            valRef.Ts = NowMs();
+
+            return valRef.Ctx;
+        }
+    }
+
+    /// <summary>
+    /// Un-deploys the specified deployment unit paths and cleans up associated job load contexts.
+    /// </summary>
+    /// <param name="deploymentUnitPaths">Deployment unit paths to undeploy.</param>
+    /// <returns><see cref="Task"/> representing the asynchronous operation.</returns>
+    public async Task<bool> UndeployUnits(ICollection<string> deploymentUnitPaths)
+    {
+        await _cacheLock.WaitAsync().ConfigureAwait(false);
+
+        try
+        {
+            bool res = false;
+
+            foreach (var deploymentUnitPath in deploymentUnitPaths)
+            {
+                if (!_deploymentUnitSets.TryGetValue(deploymentUnitPath, out var unitSet))
+                {
+                    continue;
+                }
+
+                foreach (DeploymentUnitPaths paths in unitSet)
+                {
+                    if (_jobLoadContextCache.TryGetValue(paths, out var cachedJobCtx))
+                    {
+                        cachedJobCtx.Ctx.Dispose();
+                        _jobLoadContextCache.Remove(paths);
+                    }
+                }
+
+                _deploymentUnitSets.Remove(deploymentUnitPath);
+                res = true;
+            }
+
+            return res;
+        }
+        finally
+        {
+            _cacheLock.Release();
+        }
+    }
+
+    /// <inheritdoc/>
+    public void Dispose()
+    {
+        _cts.Cancel();
+
+        _cacheLock.Wait();
+
+        try
+        {
+            foreach (var cachedJobCtx in _jobLoadContextCache)
+            {
+                cachedJobCtx.Value.Ctx.Dispose();
+            }
+
+            _jobLoadContextCache.Clear();
+            _deploymentUnitSets.Clear();
+        }
+        finally
+        {
+            _cacheLock.Release();
+            _cacheLock.Dispose();
+            _cts.Dispose();
+        }
+    }
+
+    private static long NowMs() => Stopwatch.GetTimestamp() / TicksPerMs;
```

Comment:
Why not just use `Stopwatch.GetTimestamp()`?

Reply:
We need millis to handle `ttlMs` and `cacheCleanupIntervalMs` correctly.

Reply:
Maybe reverse? i.e. calculate in the other direction?

Reply:
Good point, fixed.